[AutoDiff] Remap types more consistently during AdjointGen.

dan-zheng opened the pull request:
Remaps archetypes from original function's generic signature to
adjoint's generic signature.

Fixes crashers exposed in high-level API library.
Not adding tests to save time - compilation of the high-level API library
will soon act as a test.

dan-zheng committed:
[AutoDiff] Remap types more consistently during AdjointGen.

Remaps archetypes from original function's generic signature to
adjoint's generic signature.

Fixes crashers exposed in high-level API library.
Not adding tests to save time; compilation of the high-level API library
will soon act as a test.

dan-zheng commented:
@swift-ci Please test tensorflow macOS

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3419,6 +3430,7 @@ void AdjointEmitter::materializeZeroIndirect(CanType type,
                                              SILValue bufferAccess,
                                              SILLocation loc) {
   // Lookup `AdditiveArithmetic.zero.getter`.
+  type = remapASTType(type);
```
`materializeZeroIndirect` should in principle assume `type` has been processed. Did only substituting in visitors not solve your problem?

dan-zheng replied:
Ah, I came from a different perspective, which is to put the type remapping logic in the `materializeZero` functions to lessen the burden of putting it in all the visitors.

I also see your perspective: `materializeZero` functions should care only about "materializing zero", not remapping types.

What do you think?

rxwei replied:
Neither seems to be the best approach.  There's another way of thinking: `materializeAdjoint` is always about materializing `AdjointValue`s. The simplest approach should be to transform the type when creating an `AdjointValue` (in its c'tor and static factory methods). This makes sure all visitors will get the values of the right type because they only call `getAdjointValue`.

dan-zheng committed:
[AutoDiff] Move `AdjointValue` factory methods to `AdjointEmitter`.

During AdjointGen, archetypes from the original function must be mapped
into the adjoint function's context. To do this, it is useful to define
perform type remapping in `AdjointVAlue` factor methods defined on
`AdjointEmitter`.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2557,11 +2557,14 @@ bool AdjointGen::run() {
 /// 1, in addition to a materialized SILValue. This is expected to be passed
 /// around by value in most cases, as it's two words long.
 class AdjointValue {
+  friend class AdjointEmitter;
```
Also, make constructors be `protected`.

dan-zheng replied:
`AdjointValue`'s constructor is currently `private`, so I left it as such (since it's stricter than `protected`).

rxwei replied:
They are obviously still public in the source code.

dan-zheng replied:
Oops, done in https://github.com/apple/swift/pull/22080/commits/c17276c2b6f8c639f81ea6a218f322245afa85fc.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2755,6 +2740,13 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
         adjointGen(adjointGen), builder(getAdjoint()) {}
 
 private:
+  AdjointValue getZeroAdjointValue(SILType type);
```
I feel these should be called `make____AdjointValue`, because 1) it conflicts with the semantics of `getAdjointValue`, and 2) it's not _getting_ any intrinsic property from the emitter, but rather creating other data structures based on the emitter's context.

dan-zheng replied:
Ah, good point. Done in https://github.com/apple/swift/pull/22080/commits/e3c8cb65ccd88c1533e1822f60ae15e60beb0b57

dan-zheng committed:
Prune superfluous `remapType` calls.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3196,7 +3198,7 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
       // where `key'` is the field in the cotangent space corresponding to
       // `key`.
-      auto structTy = sei->getOperand()->getType().getASTType();
+      auto structTy = remapType(sei->getOperand()->getType()).getASTType();
```
Move line `auto av = getAdjointValue(sei);` above this and use `av.getASTType()` instead of remapping.

rxwei replied:
Actually, nvm. I think you still need a remap here.

rxwei replied:
Actually, nvm. I think you still need a remap here.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3632,7 +3649,7 @@ SILValue AdjointEmitter::accumulateMaterializedAdjointsDirect(SILValue lhs,
   assert(lhs->getType().isObject() && rhs->getType().isObject() &&
          "Adjoint types must be both object types!");
-  auto adjointTy = lhs->getType();
+  auto adjointTy = remapType(lhs->getType());
```
Since `lhs` and `rhs` are both expected to be adjoints, why is `remapType` necessary?

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/22080/commits/e3c8cb65ccd88c1533e1822f60ae15e60beb0b57.

rxwei reviewed (approved):
Great!

dan-zheng committed:
Address feedback from @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng committed:
Change `AdjointValue` constructor access level to `protected`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
@swift-ci Please test tensorflow macOS
